core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1903,13 +1903,20 @@ private[spark] class DAGScheduler(
 
           case smt: ShuffleMapTask =>
             val shuffleStage = stage.asInstanceOf[ShuffleMapStage]
-            shuffleStage.pendingPartitions -= task.partitionId
+            // Ignore task completion for old attempt of indeterminate stage
+            val ignoreIndeterminate = stage.isIndeterminate &&
+              task.stageAttemptId < stage.latestInfo.attemptNumber()
+            if (!ignoreIndeterminate) {
+              shuffleStage.pendingPartitions -= task.partitionId
+            }
             val status = event.result.asInstanceOf[MapStatus]
             val execId = status.location.executorId
             logDebug("ShuffleMapTask finished on " + execId)
             if (executorFailureEpoch.contains(execId) &&
                 smt.epoch <= executorFailureEpoch(execId)) {
               logInfo(s"Ignoring possibly bogus $smt completion from executor $execId")
```

Comment:
not related to this PR, just noticed that we may mark the partition as completed but do not register the map output.

Reply:
Agree, I am unsure about this.
+CC @Ngone51, @attilapiros who reviewed the original change - thoughts ?

Reply:
I think this explains: https://github.com/apache/spark/pull/21976/files#diff-85de35b2e85646ed499c545a3be1cd3ffd525a88aae835a9c621f877eebadcb6L1385

We should probably add back that comment.

Reply:
So this is kind of funny - take a look at what the above replaced @cloud-fan : https://github.com/apache/spark/pull/16620/files#diff-85de35b2e85646ed499c545a3be1cd3ffd525a88aae835a9c621f877eebadcb6R1183 :-)

Both of these actually do not account for DETERMINATE/INDETERMINATE changes we made subsequently.
IMO, for INDETERMINATE stages, we should ignore task completion events from previous attempts - since we have already cancelled the state.
Having said that, I have not thought through the nuances here.

+CC @jiangxb1987 as well

Reply:
I also noticed this but seems to be fine since the missing partitions are defined by `mapOutputTrackerMaster.findMissingPartitions` rather than `pendingPartitions`.

Reply:
If a task is finished but the map output is not registered, the missing partitions would be recomputed at the end of the stage, so this should be fine.